@ekino/config 2.0

tduyng opened the pull request:
## Summary
Rewrite @ekino/config for v2.0

## Motivation
- It's been so long time the library is not maintained
- Rewrite this library for adapting the modern Node.js (Typescript, ESM)

## What has changed?
Here are the key changes made in this PR:
- Rewritten in ESM and support for dual package (CommonJS and ESM)
- Removed lodash and other unnecessary dependencies
- Replaced yarn with pnpm, which is faster and more convenient with different commands and management
- Update GitHub Actions and use git-cliff on GitHub Actions for easily customizable CHANGELOG
- Replaced ESLint + Prettier with Biome, which is faster, has sufficient rules, and is easier to set up
- Rewrote all tests from jest + nyc  to vitest (compatible with ESM)

## Test plan
- All unit tests have been remained the same and fix correctly with Vitest syntax

Also close issue #40 


For the NPM release, now we use: "Trusted publishing for npm packages"
https://docs.npmjs.com/trusted-publishers#supported-cicd-providers

tduyng committed:
chore(2.0): upgrade to yarn 4.5

tduyng committed:
chore(2.0): remove husky, use native githooks

tduyng committed:
chore(2.0): upgrade prettier,remove plugin prettier eslint, reformat all files

tduyng committed:
chore(2.0): upgrade conventional-changelog-cli 5

tduyng committed:
chore(2.0): upgrade eslint 9.x and use flat config

tduyng committed:
chore(2.0): use biome, remove eslint + prettier

tduyng committed:
chore(2.0): move index to src folder for better management project

tduyng committed:
chore(2.0): simplify npm by using files fields, remove .npmignore

tduyng committed:
feat(2.0): use best practice for tsconfig.json

tduyng committed:
chore(2.0): active lint-check before creating a commit

tduyng committed:
chore: remplace yarn by pnpm

tduyng committed:
chore: setup 18.x as minimal nodejs

tduyng committed:
chore(tsconfig): add more options

tduyng committed:
chore(deps): use exact version for dev dependencies

tduyng committed:
chore(tsconfig): fix missings extends issues for tsconfig lib

tduyng committed:
feat: remove lodash, write custom native methods

tduyng committed:
feat: convert to ESM

tduyng committed:
feat: support dual package common.js and ESM

tduyng committed:
test: convert jest to vitest

tduyng committed:
fix: give correct behavior for getValue, setValue, mergeWith methods

tduyng committed:
chore(hook): use pnpm to run command

tduyng committed:
chore(utils): remove unncessary methods

tduyng committed:
test: complete all unit tests for utils methods

tduyng committed:
ci: update CI systems, run by pnpm

tduyng committed:
chore: generate changelog by git-cliff

tduyng committed:
chore: add issue templates

tduyng committed:
chore: remove deprecated links: travis CI, prettier

tduyng committed:
chore: remove yarn.lock

tduyng committed:
chore: rename global types to @types

tduyng committed:
chore: rename lint-check command to check

tduyng committed:
test: remove unused import

tduyng committed:
chore: add contributing rules

tduyng committed:
chore(deps): update latest dev dependencies

tduyng commented:
TODO: fix github-action for release and changelog.

Nico385412 commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,58 @@
+
+    strategy:
+      matrix:
+        node-version: [18.x, 20.x, 22.x]
```
```suggestion
        node-version: [20.x, 22.x, 24.x]
```

tduyng replied:
Yeah, it’s been quite a long time since this PR was created.

Nico385412 commented on .github/workflows/release.yml:
```diff
@@ -1,51 +1,88 @@
   release:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-24.04
+    if: ${{ github.actor == 'pebie' || github.actor == 'leguellec' }}
```
are you sure about that ? xD

tduyng replied:
I took the idea from other projects like veggies and node-logger. But now I’ve set it up for all maintainers and admins, not just Mika or Julos.

Nico385412 commented on .github/workflows/release.yml:
```diff
@@ -1,51 +1,88 @@
+          version: 10
+          run_install: false
+
+      - name: Setup Node.js 22.x
```
```suggestion
      - name: Setup Node.js 24.x
```

tduyng replied:
Done!

Nico385412 commented on .github/workflows/release.yml:
```diff
@@ -1,51 +1,88 @@
+      - name: Setup Node.js 22.x
+        uses: actions/setup-node@v4
+        with:
+          node-version: '22.x'
```
```suggestion
          node-version: '24.x'
```

tduyng replied:
Done!

Nico385412 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,83 @@
+
+## How to contribute
+
+Follow this guide to ensure a smooth contribution process. If you're planning significant work or substantial changes, please create an issue labeled "contribution enquiry" to discuss your proposal. This will help ensure that your work aligns with the project's goals.
```
1. The label does not exist
2. New comers cannot add labels themself

tduyng replied:
Yes, same issues like node-logger. Thank @Nico385412. Done!

tduyng committed:
feat: support minimum node.js version to 20

tduyng committed:
docs: update contributing

tduyng committed:
ci: use node 24 version to run release

tduyng committed:
test: complete utils tests for improve coverage and show coverage summary

tduyng committed:
chore(changelog): remove new contributors parts that is not correct

atacanjacques commented on README.md:
You can change for pnpm there

tduyng replied:
yes, done!

atacanjacques commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,55 @@
+    validations:
+      required: true
+    attributes:
+      label: Version of @ekino/@ekino/config
```
```suggestion
      label: Version of @ekino/config
```

tduyng replied:
Oops, thank @atacanjacques. Fixed!

atacanjacques commented on package.json:
```diff
@@ -22,44 +38,37 @@
-        "lint-check": "eslint ./*.ts ./test/*.js"
+        "lint": "biome check --write",
+        "check": "biome check",
+        "postinstall": "git config core.hooksPath ./.githooks || true"
```
Since this modify the user git config for the repo, maybe we should add this script as a step on the readme instead of running it automatically ?
What do you think ?

tduyng replied:
Yes, I add some words about that on README

halimatou-cpu reviewed (approved):
👌

tduyng committed:
fix: correct dead guard in unsetValue null-traversal check

isNullsy returns a boolean; 'boolean == null' is always false in JS,
making the left side of the || permanently dead. The null-traversal
path was only protected by the 'typeof current !== object' branch,
which silently misses an explicit null value assigned mid-path.

Replace '== null' with direct boolean evaluation so both null and
non-object intermediaries are caught correctly.

Adds a regression test for null intermediate node traversal.

tduyng committed:
fix: align cast signature in Internals type with its implementation

The Internals type declared cast as accepting an optional boolean value,
but the implementation only handles string | number as input (it produces
boolean output, it does not consume it). The mismatch let TypeScript
silently accept boolean env vars at the call site despite the runtime
function never handling that case.

- Remove boolean and the optional marker from the Internals cast type
- Narrow envVal to string | number at the call site (ProcessEnv index
  signature includes boolean for test assignment ergonomics, not runtime)

tduyng committed:
fix: replace unsafe cast in isAdvancedConfig with typeof check

The previous guard cast the input to its target type before checking,
which means an object with key explicitly set to undefined would pass
the !== undefined test and be misidentified as an advanced config.

Using typeof .key === 'string' checks the runtime shape without
assuming the cast is safe, making the guard correct by construction.

tduyng committed:
fix: capture internals.merge return value instead of relying on mutation

mergeWithDeep mutates the target when it is an object, which made the
call sites work in practice. However the return value is the canonical
result and discarding it creates a hidden coupling to the internal
mutation behaviour.

Assigning the return value back to internals.cfg makes the data-flow
explicit and correct even if the target were ever a wrapped primitive,
where the mutation would have been silently lost.

tduyng committed:
chore: update dependencies

tduyng committed:
fix: add explicit types field for TypeScript 6 compatibility

TypeScript 6 removed implicit typeRoots auto-inclusion. Previously, listing
'./node_modules/@types' in typeRoots was enough to pull in @types/node globally.
TS6 requires an explicit types: ['node'] entry in compilerOptions.

tduyng committed:
chore: clean up tsconfig for TypeScript 6 defaults and deprecations

- Remove allowSyntheticDefaultImports, esModuleInterop, strict, noImplicitAny:
  all now default to true/enabled in TS6, explicit values are redundant
- Remove typeRoots: superseded by explicit types:[node] added previously
- Add rootDir:./src to tsconfig.lib.json: TS6 no longer infers rootDir from
  input files; without it tsc errors (TS5011) and would emit to lib/src/ instead of lib/

tduyng committed:
ci: use trusted publisher OIDC based flow for release npm

tduyng committed:
ci: remove incorrect syntax, enforce strict admin users for npm release

tduyng committed:
feat: stop supporting node20 (EOL)
